source/server/filter_chain_manager_impl.cc

```diff
@@ -23,19 +32,83 @@ bool FilterChainManagerImpl::isWildcardServerName(const std::string& name) {
-                                                              std::move(filters_factory));
-  addFilterChainForDestinationPorts(destination_ports_map_, destination_port, destination_ips,
-                                    server_names, transport_protocol, application_protocols,
-                                    source_type, source_ips, source_ports, filter_chain);
+    absl::Span<const ::envoy::api::v2::listener::FilterChain* const> filter_chain_span,
+    FilterChainFactoryBuilder& b) {
+  std::unordered_set<envoy::api::v2::listener::FilterChainMatch, MessageUtil, MessageUtil>
+      filter_chains;
+  for (const auto& filter_chain : filter_chain_span) {
+    const auto& filter_chain_match = filter_chain->filter_chain_match();
+    if (!filter_chain_match.address_suffix().empty() || filter_chain_match.has_suffix_len()) {
+      throw EnvoyException(fmt::format("error adding listener '{}': contains filter chains with "
+                                       "unimplemented fields",
+                                       address_->asString()));
+    }
+    if (filter_chains.find(filter_chain_match) != filter_chains.end()) {
+      throw EnvoyException(fmt::format("error adding listener '{}': multiple filter chains with "
+                                       "the same matching rules are defined",
+                                       address_->asString()));
+    }
+    filter_chains.insert(filter_chain_match);
+
+    // If the cluster doesn't have transport socket configured, then use the default "raw_buffer"
+    // transport socket or BoringSSL-based "tls" transport socket if TLS settings are configured.
+    // We copy by value first then override if necessary.
+    auto transport_socket = filter_chain->transport_socket();
+    if (!filter_chain->has_transport_socket()) {
+      if (filter_chain->has_tls_context()) {
+        transport_socket.set_name(Extensions::TransportSockets::TransportSocketNames::get().Tls);
+        MessageUtil::jsonConvert(filter_chain->tls_context(), *transport_socket.mutable_config());
+      } else {
+        transport_socket.set_name(
+            Extensions::TransportSockets::TransportSocketNames::get().RawBuffer);
+      }
+    }
+
+    auto& config_factory = Config::Utility::getAndCheckFactory<
+        Server::Configuration::DownstreamTransportSocketConfigFactory>(transport_socket.name());
+    ProtobufTypes::MessagePtr message = Config::Utility::translateToFactoryConfig(
+        transport_socket, validation_visitor_, config_factory);
+
+    // Validate IP addresses.
+    std::vector<std::string> destination_ips;
+    destination_ips.reserve(filter_chain_match.prefix_ranges().size());
+    for (const auto& destination_ip : filter_chain_match.prefix_ranges()) {
+      const auto& cidr_range = Network::Address::CidrRange::create(destination_ip);
+      destination_ips.push_back(cidr_range.asString());
+    }
+
+    std::vector<std::string> server_names(filter_chain_match.server_names().begin(),
```

Comment:
Since filter chain creation is now happening in `ListenerFilterChainFactoryBuilder::buildFilterChain()`, we don't need to create `server_names` here, instead, you can just loop over `filter_chain_match.server_names()` in the code below.

Reply:
Probably not. 
1. buildFilterChain need a copy of server names.
2. addFilterChainForDestinationPorts currently need a vector as well
3. for loop may not need a vector but 2) need.
I mark 2 as TODO but yet the vector is needed for now

Reply:
Ah, right, I've missed `addFilterChainForDestinationPorts()`.